src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -784,78 +794,132 @@ object Gen extends GenArities with GenVersionSpecific {
 
   //// Character Generators ////
 
+  private def charSample(cs: Array[Char]): Gen[Char] =
+    new Gen[Char] {
+      def doApply(p: P, seed0: Seed): Gen.R[Char] = {
+        val (x, seed1) = seed0.long
+        val i = ((x & Long.MaxValue) % cs.length).toInt
+        r(Some(cs(i)), seed1)
+      }
+    }
+
   /** Generates a numerical character */
-  def numChar: Gen[Char] = choose(48.toChar, 57.toChar)
+  val numChar: Gen[Char] =
+    charSample(('0' to '9').toArray)
 
   /** Generates an upper-case alpha character */
-  def alphaUpperChar: Gen[Char] = choose(65.toChar, 90.toChar)
+  val alphaUpperChar: Gen[Char] =
+    charSample(('A' to 'Z').toArray)
 
   /** Generates a lower-case alpha character */
-  def alphaLowerChar: Gen[Char] = choose(97.toChar, 122.toChar)
+  val alphaLowerChar: Gen[Char] =
+    charSample(('a' to 'z').toArray)
 
   /** Generates an alpha character */
-  def alphaChar = frequency((1,alphaUpperChar), (9,alphaLowerChar))
+  val alphaChar: Gen[Char] =
+    charSample((('A' to 'Z') ++ ('a' to 'z')).toArray)
 
   /** Generates an alphanumerical character */
-  def alphaNumChar = frequency((1,numChar), (9,alphaChar))
+  val alphaNumChar: Gen[Char] =
+    charSample((('0' to '9') ++ ('A' to 'Z') ++ ('a' to 'z')).toArray)
 
   /** Generates a ASCII character, with extra weighting for printable characters */
-  def asciiChar: Gen[Char] = chooseNum(0, 127, 32 to 126:_*).map(_.toChar)
+  val asciiChar: Gen[Char] =
+    choose(0, 127).map(_.toChar)
```

Comment:
why a different strategy here than say `charSample`?

Reply:
I think `charSample` is best when you need to allocate a bunch of non-continuous `Char` ranges that would require a lot of branching to support. For `asciiChar` and `asciiPrintableChar` there's a single contiguous range.

That said, since each only uses around 100 characters, there's no real problem if we did want to allocate a static array to sample from.